ibc/store/proofs_ics23.go

```diff
@@ -0,0 +1,159 @@
+func createNonMembershipProof(tree *smt.SMT, key []byte) (*ics23.CommitmentProof, error) {
+	proof, err := tree.Prove(key)
+	if err != nil {
+		return nil, coreTypes.ErrCreatingProof(err)
+	}
+
+	return convertSMPToExclusionProof(&proof, key), nil
+}
+
+// convertSMPToExistenceProof converts a SparseMerkleProof to an ics23
+// ExistenceProof to verify membership of an element
+func convertSMPToExistenceProof(proof *smt.SparseMerkleProof, key, value []byte) *ics23.CommitmentProof {
+	path := sha256.Sum256(key)
+	steps := convertSideNodesToSteps(proof.SideNodes, path[:])
+	return &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Exist{
+			Exist: &ics23.ExistenceProof{
+				Key:   key,
+				Value: value,
+				Leaf:  smtSpec.LeafSpec,
+				Path:  steps,
+			},
+		},
+	}
+}
+
+// convertSMPToExistenceProof converts a SparseMerkleProof to an ics23
+// ExclusionProof to verify non-membership of an element
+func convertSMPToExclusionProof(proof *smt.SparseMerkleProof, key []byte) *ics23.CommitmentProof {
+	path := sha256.Sum256(key)
+	steps := convertSideNodesToSteps(proof.SideNodes, path[:])
+	leaf := &ics23.LeafOp{
+		Hash: ics23.HashOp_SHA256,
+		// Do not re-hash already hashed fields from NonMembershipLeafData
+		PrehashKey:   ics23.HashOp_NO_HASH,
+		PrehashValue: ics23.HashOp_NO_HASH,
+		Length:       ics23.LengthOp_NO_PREFIX,
+		Prefix:       []byte{0},
+	}
+	actualPath := path[:]
+	actualValue := defaultValue
+	if proof.NonMembershipLeafData != nil {
+		actualPath = proof.NonMembershipLeafData[1:33]
+		actualValue = proof.NonMembershipLeafData[33:]
+	}
+	return &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Exclusion{
+			Exclusion: &ics23.ExclusionProof{
+				Key:             key,
+				ActualPath:      actualPath,
+				ActualValueHash: actualValue,
+				Leaf:            leaf,
+				Path:            steps,
+			},
+		},
+	}
+}
+
+// convertSideNodesToSteps converts the SideNodes field in the SparseMerkleProof
+// into a list of InnerOps for the ics23 CommitmentProof
+func convertSideNodesToSteps(sideNodes [][]byte, path []byte) []*ics23.InnerOp {
+	steps := make([]*ics23.InnerOp, 0, len(sideNodes))
+	for i := 0; i < len(sideNodes); i++ {
+		var prefix, suffix []byte
+		prefix = append(prefix, innerPrefix...)
+		if getPathBit(path, len(sideNodes)-1-i) == left {
+			suffix = make([]byte, 0, len(sideNodes[i]))
+			suffix = append(suffix, sideNodes[i]...)
+		} else {
+			prefix = append(prefix, sideNodes[i]...)
+		}
+		op := &ics23.InnerOp{
+			Hash:   ics23.HashOp_SHA256,
+			Prefix: prefix,
+			Suffix: suffix,
+		}
+		steps = append(steps, op)
+	}
+	return steps
+}
+
+// getPathBit determines whether the hash of a node at a certain depth in the tree is the
+// left or the right child of its parent
+func getPathBit(data []byte, position int) position {
+	if int(data[position/8])&(1<<(8-1-uint(position)%8)) > 0 {
```

Comment:
Can you provide an example of what this does so others avoid spending the time figuring out what you already have